File: id_utils.py
"""
id_utils.py
Deterministyczne ID pytań - lustrzane odbicie id-utils.js.
Używane przeze mnie offline przy migracji starych baz (anatomia/histologia/biochemia)
do nowego wspólnego schematu, żeby ID zgadzały się z tym, co appka policzy w przeglądarce.
"""
import re

SUBJECT_PREFIX = {
    "anatomia": "anat",
    "histologia": "hist",
    "biochemia": "bioch",
    "ebm": "ebm",
    "historia_medycyny": "hmed",
    "angielski_2": "ang2",
    "mikrobiologia": "mikro",
}


def normalize_text(s: str) -> str:
    s = re.sub(r'^\s*\d+\.\s*', '', s)  # usuń numerację "123. " z początku
    s = s.strip()
    s = re.sub(r'\s+', ' ', s)
    return s


def fnv1a(s: str) -> str:
    """FNV-1a 32-bit nad bajtami UTF-8 - identyczne z wersją JS (TextEncoder)."""
    h = 0x811c9dc5
    for byte in s.encode('utf-8'):
        h ^= byte
        h = (h * 0x01000193) & 0xFFFFFFFF
    return format(h, '08x')


def make_question_id(subject: str, q: str, o: list[str]) -> str:
    prefix = SUBJECT_PREFIX.get(subject, subject[:5])
    norm_q = normalize_text(q)
    norm_o = "|".join(x.strip() for x in o)
    h = fnv1a(norm_q + "||" + norm_o)
    return f"{prefix}_{h}"


def assign_ids(questions: list[dict], subject: str) -> list[dict]:
    out = []
    for item in questions:
        if not item.get("id"):
            item = dict(item)
            item["id"] = make_question_id(subject, item["q"], item["o"])
        out.append(item)
    return out


if __name__ == "__main__":
    # szybki self-test
    sample_q = "1. Gdzie znajduje się jądro wierzchu (nucleus fastigii)?"
    sample_o = ["Wzgórze", "Most", "Móżdżek", "Śródmózgowie", "Rdzeń przedłużony"]
    print(make_question_id("anatomia", sample_q, sample_o))

File: id-utils.js
// id-utils.js
// Deterministyczne ID pytań. Ten sam algorytm co w id_utils.py (Python),
// żeby ID liczone offline (przy migracji bazy) i w przeglądarce się zgadzały.

const SUBJECT_PREFIX = {
    anatomia: "anat",
    histologia: "hist",
    biochemia: "bioch",
    ebm: "ebm",
    historia_medycyny: "hmed",
    angielski_2: "ang2",
    mikrobiologia: "mikro",
};

function normalizeText(s) {
    return s
        .replace(/^\s*\d+\.\s*/, "")   // usuń numerację "123. " z początku
        .trim()
        .replace(/\s+/g, " ");
}

// FNV-1a 32-bit nad bajtami UTF-8 (TextEncoder), żeby wynik był identyczny
// z Pythonem (s.encode('utf-8')) niezależnie od polskich znaków diakrytycznych.
function fnv1a(str) {
    const bytes = new TextEncoder().encode(str);
    let h = 0x811c9dc5;
    for (let i = 0; i < bytes.length; i++) {
        h ^= bytes[i];
        h = Math.imul(h, 0x01000193) >>> 0;
    }
    return (h >>> 0).toString(16).padStart(8, "0");
}

function makeQuestionId(subject, q, o) {
    const prefix = SUBJECT_PREFIX[subject] || subject.slice(0, 5);
    const normQ = normalizeText(q);
    const normO = o.map(x => x.trim()).join("|");
    const hash = fnv1a(normQ + "||" + normO);
    return `${prefix}_${hash}`;
}

// Dołącza `id` do każdego pytania w tablicy, jeśli go jeszcze nie ma.
function assignIds(questions, subject) {
    return questions.map(q => ({
        ...q,
        id: q.id || makeQuestionId(subject, q.q, q.o),
    }));
}

if (typeof module !== "undefined") {
    module.exports = { makeQuestionId, assignIds, normalizeText, fnv1a };
}
